routes/v0/cases.js

```diff
@@ -5,6 +5,25 @@ const { Contact, Case, CaseAudit, sequelize } = models;
 const ContactController = require('../../controllers/contact-controller')(Contact);
 const CaseController = require('../../controllers/case-controller')(Case, sequelize);
 const CaseAuditController = require('../../controllers/case-audit-controller')(CaseAudit);
+const { CaseResource, can } = require('../../permissions');
+const { asyncHandler, unauthorized } = require('../../utils');
+
+/**
+ * This methods checks if the user can edit the case.
+ * If yes, the endpoint code is run.
+ * If not, it sends unauthorized.
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ */
+const editCasePermissions = async (req, res, next) => {
```

Comment:
For readability, you could consider `const creatorCanEditCase = async (req...)`.  Then in the function that handles the endpoint, you'd have:
```
casesRouter.put('/:id', asyncHandler(creatorCanEditCase), async (req, res) => {
```
It might be good for readability.

I wonder if it would be possible to do a series of ORs, like so:
```
casesRouter.put('/:id', asyncHandler(creatorCanEditCase), asyncHandler(supervisorCanEditCase), async (req, res) => {
```
Any of those have the opportunity to set a bit to `true` and let it continue.  You may need another middleware at the end to block it:
```
casesRouter.put('/:id', asyncHandler(creatorCanEditCase), asyncHandler(supervisorCanEditCase), blockUnauthorized, async (req, res) => {
```
Maybe `blockUnauthorized` could be added by default for all requests somehow.

Or maybe this... you have a middleware for all requests that goes at the very start of the chain that adds something like:
```
req.authorized = false
```
And you have `blockUnauthorized` added for all requests at the end of the chain, which just does:
```
if !(req.authorized) { unauthorized();}
else next();
```
And then when you define the endpoint, you can add middleware in between those two in order to allow `cancan` to set `authorized = true`.

I don't know if that would work, but it seems like an interesting idea.

Reply:
I've just implemented `SafeRouter` that works just like this.